### PR TITLE
add metrics to OperationPool

### PR DESCRIPTION
**Pull request by seroze**
## PR Description
Fix for issue #3917 (Metrics for operation pool size and usage), creating a new PR because the old one(https://github.com/ConsenSys/teku/pull/4295) had issue with CLA signing 
## Documentation

- [ ] I thought about documentation and added the `documentation` label to this PR if updates are required.

## Changelog

- [ ] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

**Comment by seroze** [comment]
@ajsutton  I have created a new PR because of the CLA issue(have closed the old one), I tried to experiment with the git rebase command you have suggested, but it went haywire  . 

> This may be causing issues - track validation reason won't be a valid metric label name because of the spaces. I'd suggest just result.

done

>I'd probably name these operation_pool_size_ and operation_pool_validation_

done

**Comment by ajsutton** [comment]
New PR is looking good.  The AT is failing because the voluntary exit is getting:
```
2021-09-02 08:36:49.665+10:00 | docker-java-stream--1146902259 | DEBUG | TekuVoluntaryExit | Invalid params response from Beacon Node API (url = http://tekunode3:9051/eth/v1/beacon/pool/voluntary_exits, response = {"status":400,"message":"Incorrect number of labels."})
```

The metric definitions are missing the "help" text so the label name is being used as help instead of the label, leading to this error.  I'll leave a suggested change to make that a bit clearer.

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationPool.java [hunk 1]
```diff
@@ -31,16 +35,29 @@
...
     this.slotToSszListSchemaSupplier = slotToSszListSchemaSupplier;
     this.operationValidator = operationValidator;
+
+    metricsSystem.createIntegerGauge(
+        TekuMetricCategory.BEACON, OPERATION_POOL_SIZE_METRIC + metricType, "result", this::size);
+    validationReasonCounter =
+        metricsSystem.createLabelledCounter(
+            TekuMetricCategory.BEACON,
+            OPERATION_POOL_SIZE_VALIDATION_REASON + metricType,
+            "result");
```
Comment (ajsutton) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationPool.java]:
```suggestion
            OPERATION_POOL_SIZE_VALIDATION_REASON + metricType,
            "Total number of attempts to add an operation to the pool, broken down by validation result",
            "result");
```

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationPool.java [hunk 2]
```diff
@@ -31,16 +35,29 @@
...
   public OperationPool(
+      String metricType,
+      MetricsSystem metricsSystem,
       Function<UInt64, SszListSchema<T, ?>> slotToSszListSchemaSupplier,
       OperationValidator<T> operationValidator) {
     this.slotToSszListSchemaSupplier = slotToSszListSchemaSupplier;
     this.operationValidator = operationValidator;
+
+    metricsSystem.createIntegerGauge(
+        TekuMetricCategory.BEACON, OPERATION_POOL_SIZE_METRIC + metricType, "result", this::size);
```
Comment (ajsutton) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationPool.java]:
```suggestion
        TekuMetricCategory.BEACON, 
        OPERATION_POOL_SIZE_METRIC + metricType, 
        "Current number of operations in the pool", 
        this::size);
```

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationPool.java (hunk 1 above)
Comment (seroze) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationPool.java]: done, thanks for this suggestion,  my only concern is if we need to pass non-zero number of labels, metrics contract should have enforced this somehow.

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationPool.java (hunk 2 above)
Comment (seroze) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationPool.java]: done

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationPool.java (hunk 1 above)
Comment (ajsutton) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/OperationPool.java]: It doesn't have to be non-zero, just that the number of label names has to match the number of label values provided.  Unfortunately there isn't a good way of enforcing that at compile time while still supporting an arbitrary number of labels (and we ultimately inherit that API from Prometheus' library).

**Review by ajsutton — APPROVED** [review]
LGTM. Thanks for taking this on and your patience in dealing with the CLA bot. :)

**Comment by seroze** [comment]
> New PR is looking good. The AT is failing because the voluntary exit is getting:
> 
> ```
> 2021-09-02 08:36:49.665+10:00 | docker-java-stream--1146902259 | DEBUG | TekuVoluntaryExit | Invalid params response from Beacon Node API (url = http://tekunode3:9051/eth/v1/beacon/pool/voluntary_exits, response = {"status":400,"message":"Incorrect number of labels."})
> ```
> 
> The metric definitions are missing the "help" text so the label name is being used as help instead of the label, leading to this error. I'll leave a suggested change to make that a bit clearer.

how did you find this specific  log, I couldn't see it in here [https://app.circleci.com/pipelines/github/ConsenSys/teku/13374/workflows/8f3b069a-d19e-4747-9f68-35c57a425841/jobs/87413](url) , did you run it locally and figure this out

**Comment by ajsutton** [comment]
> how did you find this specific log, I couldn't see it in here [https://app.circleci.com/pipelines/github/ConsenSys/teku/13374/workflows/8f3b069a-d19e-4747-9f68-35c57a425841/jobs/87413](url) , did you run it locally and figure this out


Yeah I had to run it locally to see it.  I think the test artefacts do capture a bunch of additional information and may have the full log output for tests, but it's so much easier to debug stuff when you can reproduce it locally.

**Pull request merged**